src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/BannedTypesAnalyzer.cs

```diff
@@ -0,0 +1,91 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Azure.ClientSdk.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class BannedTypesAnalyzer : SymbolAnalyzerBase
+    {
+        private static HashSet<string> BannedTypes = new HashSet<string>()
+        {
+            "MutableJsonDocument",
+            "MutableJsonElement",
+            "MutableJsonChange"
+        };
+
+        private static readonly string BannedTypesMessageArgs = string.Join(", ", BannedTypes);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Descriptors.AZC0020);
+
+        public override SymbolKind[] SymbolKinds { get; } = new[]
+        {
+            SymbolKind.Method,
+            SymbolKind.Field,
+            SymbolKind.Property,
+            SymbolKind.Parameter,
+            SymbolKind.Event,
+            SymbolKind.NamedType
+        };
+
+        public override void Analyze(ISymbolAnalysisContext context)
+        {
+            static void CheckType(ISymbolAnalysisContext context, ITypeSymbol type, ISymbol symbol)
+            {
+                if (type is INamedTypeSymbol namedTypeSymbol)
+                {
+                    if (BannedTypes.Contains(type.Name))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0020, symbol.Locations.First(), BannedTypesMessageArgs), symbol);
+                    }
+
+                    if (namedTypeSymbol.IsGenericType)
+                    {
+                        foreach (var typeArgument in namedTypeSymbol.TypeArguments)
+                        {
+                            CheckType(context, typeArgument, symbol);
+                        }
+                    }
+                }
+            }
```

Comment:
IMO, this could be a little confusing given `Analyze` is also passed a `context`, which you now shadow. And because it's the only parameter you could possibly want to use in an enclosed static method, I don't see a point in not just declaring this method on the class and passing in the context, which you have to do anyway. Would perhaps make it a little more readable / maintainable.

Reply:
Cool, happy to change that.  I copied this pattern from [BannedAssembliesAnalyzer](https://github.com/Azure/azure-sdk-tools/blob/main/src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/BannedAssembliesAnalyzer.cs), but don't feel strongly about it.